src/mds/Server.cc

```diff
@@ -769,6 +769,9 @@ void Server::find_idle_sessions()
       break;
     }
 
+    if (not mds->locker->caps_reqd_by_other_client((session)->get_client()))
+      continue;
+
```

Comment:
I thought about this again.  we should apply apply the check to sessions in 'open' state. because sessions in 'stale' stale already have all theirs caps revoked.

For sessions in 'open' state, we should also check if anyone is flushing session messages (see Server::finish_flush_session)

besides, I think it's wrong to keep dead sessions infinitely. maybe just enlarge session_timeout if no one else want its caps.

Reply:
>  we should apply apply the check to sessions in 'open' state. because sessions in 'stale' stale already have all theirs caps revoked.

I thought it was just **stale** caps were revoked?

https://github.com/ceph/ceph/blob/b2a699cb737a4b4d9110db69e895d8e1f772117a/src/mds/Server.cc#L727

Reply:
All caps of stale sessions are revoked. what does un-pinned  caps means?

Reply:
> I thought about this again. we should apply apply the check to sessions in 'open' state. because sessions in 'stale' stale already have all theirs caps revoked.

Ok. Umm, but what should be done if an "open" session is on the caps revoke list? We can't evict it since it still has to pass `session_timeout` and `session_autoclose`.

Reply:
if no one else want its caps, delay moving it the to stale state

Reply:
Okay, got it. How much should the delay be? Should I add a new option for "extra timeout"? We can also keep delaying a client's eviction as long as nobody wants it caps.

And, the current code that makes "stale" client's eviction delay should stay as a different patch, right?

Reply:
>All caps of stale sessions are revoked. what does un-pinned caps means?

https://github.com/ceph/ceph/blob/e7147fb34cb6ee2830d2af2ec0efb3678b7995b9/src/mds/Locker.cc#L2142-L2160

Caps with `p` (pinned) are not revoked?